File: csharp/prime-factors/PrimeFactors.cs
using System;
using System.Collections.Generic;

public static class PrimeFactors
{
    private static bool[] _sieve = new bool[0];
    public static long[] Factors(long number)
    {
        if (number <= 1L) return new long[0];
        List<long> primes = new List<long>();
        while (number % 2L == 0)
        {
            primes.Add(2L);
            number /= 2L;
        }
        for (long factor = 3L; factor <= number; factor += 2L)
        {
            while (number % factor == 0)
            {
                primes.Add(factor);
                number /= factor;
            }
        }
        return primes.ToArray();
    }
}

File: fsharp/prime-factors/PrimeFactors.fs
module PrimeFactors

open System

let factors (number: int64) : int32 list = 
    if number <= 1L then []
    else 
        let mutable num = number
        let mutable primes = []
        while num % 2L = 0L do
            num <- num / 2L
            primes <- 2 :: primes
        for n in [3..2..(Math.Sqrt(num |> float) |> int)] do
            let k = int64 n
            while num % k = 0L do
                num <- num / k
                primes <- primes @ [n]
        if num > 1L then primes @ [num |> int]
        else
            primes
            
        